src/components/candevice/candevicemodel.cpp

```diff
@@ -0,0 +1,98 @@
+#include "candevicemodel.h"
+#include <QtCore/QDir>
+#include <QtCore/QEvent>
+
+#include <QtWidgets/QFileDialog>
+
+#include <nodes/DataModelRegistry>
+
+#include "datamodeltypes/candevicedata.h"
+
+CanDeviceModel::CanDeviceModel()
+    : label(new QLabel())
+{
+    label->setAlignment(Qt::AlignVCenter | Qt::AlignHCenter);
+
+    label->setFixedSize(75, 25);
+
+    label->installEventFilter(this);
+
+    label->setAttribute(Qt::WA_TranslucentBackground);
+
+    canDevice = new CanDevice(factory);
+
+    connect(canDevice, &CanDevice::frameSent, this, &CanDeviceModel::frameSent);
+    connect(canDevice, &CanDevice::frameReceived, this, &CanDeviceModel::frameReceived);
+
+    canDevice->init("socketcan", "can0");
+    canDevice->start();
+}
+
+CanDeviceModel::~CanDeviceModel() { delete canDevice; }
+
+unsigned int CanDeviceModel::nPorts(PortType portType) const
+{
+    unsigned int result = 1;
+
+    switch (portType) {
+    case PortType::In:
+        result = 1;
+        break;
+
+    case PortType::Out:
+        result = 1;
+        break;
+
+    case PortType::None:
+        result = 0;
+        break;
+    }
+
+    return result;
+}
+
+void CanDeviceModel::frameReceived(const QCanBusFrame& frame)
+{
+    _frame = frame;
```

Comment:
For now I have no proof if this would work with multiple threads or not. I will add frame queue, just to be sure. Better safe than sorry.